File: dart/test/models_test.dart
import "package:test/test.dart";

import "fixtures/helper.dart" as helper;

const double delta = 0.001;

void main() {
  group("Circle Tests", () {
    test("Test Circle Area", () {
      var circle = helper.setupCircle();
      var result = circle.area();
      const expected = 754.767;
      expect((result - expected).abs() < delta, true);
    }, tags: ["area"]);

    test("Test Circle Circumference", () {
      var circle = helper.setupCircle();
      var result = circle.perimeter();
      const expected = 97.389;
      expect((result - expected).abs() < delta, true);
    }, tags: ["perimeter"]);
  }, tags: ["circle"]);

  group("Rectangle Tests", () {
    test("Test Rectangle Area", () {
      var rectangle = helper.setupRectangle();
      var result = rectangle.area();
      const expected = 756.437;
      expect((result - expected).abs() < delta, true);
    }, tags: ["area"]);

    test("Test Rectangle Perimeter", () {
      var rectangle = helper.setupRectangle();
      var result = rectangle.perimeter();
      const expected = 112.0;
      expect((result - expected).abs() < delta, true);
    }, tags: ["perimeter"]);
  }, tags: ["rectangle"]);

  group("Square Tests", () {
    test("Square Area", () {
      var square = helper.setupSquare();
      var result = square.area();
      const expected = 105.062;
      expect((result - expected).abs() < delta, true);
    }, tags: ["area"]);

    test("Square Perimeter", () {
      var square = helper.setupSquare();
      var result = square.perimeter();
      const expected = 41.0;
      expect((result - expected).abs() < delta, true);
    }, tags: ["perimeter"]);
  }, tags: ["square"]);
}

File: haskell/test/ModelsTest.hs
module Main (main) where

import Test.Hspec

import Shapes.Base.Commons

import Fixtures.Helper qualified as Helper

delta :: Double
delta = 0.001

main :: IO ()
main = hspec $ do
    describe "Models Tests" $ do
        circleTests
        rectangleTests
        squareTests

circleTests :: Spec
circleTests = describe "Circle Tests" $ do
    it "Test Circle Area" $ do
        let circle = Helper.setupCircle
        let result = area circle
        let expected = 754.767
        abs (result - expected) `shouldSatisfy` (< delta)

    it "Test Circle Perimeter" $ do
        let circle = Helper.setupCircle
        let result = perimeter circle
        let expected = 97.389
        abs (result - expected) `shouldSatisfy` (< delta)

rectangleTests :: Spec
rectangleTests = describe "Rectangle Tests" $ do
    it "Test Rectangle Area" $ do
        let rectangle = Helper.setupRectangle
        let result = area rectangle
        let expected = 756.437
        abs (result - expected) `shouldSatisfy` (< delta)

    it "Test Rectangle Perimeter" $ do
        let rectangle = Helper.setupRectangle
        let result = perimeter rectangle
        let expected = 112.0
        abs (result - expected) `shouldSatisfy` (< delta)

squareTests :: Spec
squareTests = describe "Square Tests" $ do
    it "Test Square Area" $ do
        let square = Helper.setupSquare
        let result = area square
        let expected = 105.062
        abs (result - expected) `shouldSatisfy` (< delta)

    it "Test Square Perimeter" $ do
        let square = Helper.setupSquare
        let result = perimeter square
        let expected = 41.0
        abs (result - expected) `shouldSatisfy` (< delta)
